zerver/views/users.py

```diff
@@ -261,6 +262,11 @@ def add_bot_backend(request, user_profile, full_name_raw=REQ("full_name"), short
     full_name = check_full_name(full_name_raw)
     email = '%s@%s' % (short_name, user_profile.realm.get_bot_domain())
     form = CreateUserForm({'full_name': full_name, 'email': email})
+    if bot_type == UserProfile.EMBEDDED_BOT:
+        is_service_valid = any(service_name == embedded_bot_service.name for embedded_bot_service in EMBEDDED_BOTS)
+        if not is_service_valid:
+            return json_error(_("Invalid service name."))
```

Comment:
I'm not a fan of this `any` constructon.  I would spell it as:

```
if service_name not in [embedded_bot_service.name for embedded_bot_service in EMBEDDED_BOTS]:
    return json_error(_("Invalid embedded bot name."))
```

(I don't think "service" should appear in user-facing strings at all)

Otherwise, this commit looks good.

Reply:
Updated the `any` construction in the commit `users.py: Clean up and shorten bot handling code.` . Is it possible that you were referring to that commit, but made the comment on Abhijeet's original commit?

Reply:
I think probably you should create a squashed/clean up version of this backend commit closer to the bottom of the branch that we can try to merge next.

(Just so you understand strategy, I'm trying to avoid adding the UI until this works how we want it to from a security perspective).

Reply:
> Just so you understand strategy, I'm trying to avoid adding the UI until this works how we want it to from a security perspective.

Yeah, that makes sense, now shortly before the release. Do you already see any specific sec issues or is it more that you haven't got to carefully review the frontend part yet?

Reply:
It's more that backend bugs are where security issues come from.

Reply:
Will refactor then. It would come really convenient if you could merge the commit `embedded bots: Alter tests to accept any class name. ` from #7043, since that's what prevents the tests from passing. It'd be easier for me to judge any mistakes in this PR afterwards.

Reply:
Sounds good.  Just merged that :)